pylearn2/datasets/hdf5.py

```diff
@@ -2,338 +2,472 @@
-        shape = [V.shape[axes.index('b')],
-                 V.shape[axes.index(0)],
-                 V.shape[axes.index(1)],
-                 V.shape[axes.index('c')]]
-        self.view_converter = HDF5ViewConverter(shape[1:], axes=axes)
-        self.X = self.view_converter.topo_view_to_design_mat(V)
-        # self.X_topo_space stores a "default" topological space that
-        # will be used only when self.iterator is called without a
-        # data_specs, and with "topo=True", which is deprecated.
-        self.X_topo_space = self.view_converter.topo_space
-
-        # Update data specs
-        X_space = VectorSpace(dim=V.shape[axes.index('b')])
-        X_source = 'features'
-        if self.y is None:
-            space = X_space
-            source = X_source
+        data = alias_dict()
+        if use_h5py:
+            for s, a in safe_zip(sources, aliases):
+                if load_all:
+                    data[s, a] = self._fhandler[s][:]
+                else:
+                    data[s, a] = self._fhandler[s]
+                    # hdf5 handle has no ndim
+                    data[s].ndim = len(data[s].shape)
         else:
-            if self.y.ndim == 1:
-                dim = 1
-            else:
-                dim = self.y.shape[-1]
-            y_space = VectorSpace(dim=dim)
-            y_source = 'targets'
-            space = CompositeSpace((X_space, y_space))
-            source = (X_source, y_source)
-
-        self.data_specs = (space, source)
-        self.X_space = X_space
-        self._iter_data_specs = (X_space, X_source)
+            for s, a in safe_zip(sources, aliases):
+                if load_all:
+                    data[s, a](self._fhandler.getNode('/', s)[:])
+                else:
+                    data[s, a] = self._fhandler.getNode('/', s)
+        return data
 
+    @wraps(Dataset.iterator, assigned=(), updated=())
+    def iterator(self, mode=None, batch_size=None, num_batches=None,
+                 rng=None, data_specs=None, return_tuple=False,
+                 **kwargs):
+        """
+        data_specs : tuple
+            A `(space, source)` tuple. See :ref:`data_specs` for a full
+            description.
+        return_tuple : bool, optional
+            Always return a tuple, even if there is exactly one source
+            of data being returned. Defaults to `False`.
+        """
+        [mode, batch_size, num_batches, rng, data_specs] = self._init_iterator(
+            mode, batch_size, num_batches, rng, data_specs)
+        convert = None
+
+        if data_specs is None:
+            sources = self._get_canonical_sources()
+            spaces = self._get_canonical_spaces()
+            data_specs = (spaces, sources)
+            data = self._get_canonical_data()
+        else:
+            sources = data_specs[1]
+            data = [self.data[s] for s in sources]
+
+        return FiniteDatasetIterator(self,
+                                     mode(data[0].shape[0],
```

Comment:
That's right, your code currently works. I'm just pointing out that you already have that code factored out in `self.get_num_examples()`, which is also part of the `Dataset` interface, so it would be more legible to use it. Also, if `get_num_examples` gets upgraded in the future (let's say, to perform so checking and cache the value), then you would get those benefits here as well.
It's jut a minor point, though.

When the iterator is called, we iterate jointly on as many sources as needed. This is essential if we want to keep the targets associated with the right inputs, for instance.
One assumption that is made by FiniteDatasetIterator, that could be more explicit, is that all sources have the same number of elements, and that elements at a given index in different sources always belong to the same example.